### ISSUE TITLE
Queue not playing next songs and bot partially responding to commands

### ISSUE THREAD

**Issue by BackIsBachus** [matthieugrieger/mumbledj#174]
Hi,

#### Before submitting this issue, please acknowledge that you have done the following:
- [x] I've at least skimmed through the [README](https://github.com/matthieugrieger/mumbledj/blob/master/README.md)
- [x] I have checked that I am running the latest version of MumbleDJ (use `mumbledj --version` when starting the bot or use the MumbleDJ version command in Mumble)
- [x] I have [searched through the existing issues](https://github.com/matthieugrieger/mumbledj/issues) to see if my issue has been answered already

I use the latest released binary (couldn't manage to install go1.5+ so didn't use go get method) but the version command returns 3.1.1 instead of 3.1.2

---

#### What type of issue is this?
- [x] Bug report (encountered problems with MumbleDJ)
- [ ] Feature request (request for a new functionality)
- [ ] Question
- [ ] Other:

---

#### Log output of bot with `--debug` flag (likely only for bug reports):

```
INFO[0000] Checking for duplicate aliases...
INFO[0000] Performing startup checks...
INFO[0000] Checking for availability of services...      num_services=3
WARN[0000] A startup check discovered an issue. The service will be disabled.  error=No SoundCloud API key has been provided service=SoundCloud
INFO[0000] Checking YouTubeDL installation...
INFO[0000] Checking avconv installation...
INFO[0000] Checking aria2c installation...
INFO[0000] Checking openssl installation...
INFO[0000] Attempting connection to server...            address=127.0.0.1 port=64738
INFO[0002] Setting default volume...                     volume=0.20
INFO[0002] Caching disabled.
INFO[0011] Sending a message to channel...               channel=Root message=<b>BackIsBachus</b> added <b>51</b> tracks to the queue.
INFO[0103] Sending a private message...                  message=MumbleDJ version: <b>v3.1.1</b> user=BackIsBachus
INFO[0137] Sending a private message...                  message=There are currently <b>51</b> tracks in the queue. user=BackIsBachus
INFO[0226] Sending a private message...                  message=MumbleDJ version: <b>v3.1.1</b> user=BackIsBachus
```

---

#### Description of your issue:
* I add a playlist (or several individual songs) from Youtube
* The 1st song plays until it ends (or it is skipped)
* Meanwhile I ask the bot its version and the number of tracks (it responds)
* Nothing happens (ie. the next songs never plays)
* The bot still responds to some commands (joinme, move, pause, resume, cachesize, help, volume, reload, version, kill at least) but not to others (numtracks, listtracks, currenttrack, reset, shuffle, add, addnext at least). Also the commands `pause` and `resume` inform me that `Error: Either the audio is already paused, or there are no tracks in the queue.`

I have tested this with or without cache and with or without shuffle activated.

If you need any other detail I'd be happy to provide them

Thanks!
BackIsBachus

**Comment by matthieugrieger** [comment]
Have you tried updating `youtube-dl`? Sometimes this fixes issues like the one you are experiencing.

The reason why some commands are working and some aren't is because MumbleDJ is now multithreaded. Essentially each new command that the bot picks up spawns a new thread, and that thread handles the execution of that command. numtracks, listtracks, currenttrack, etc. all have one thing in common: they interact with the queue. There is a locking mechanism on core data structures in the bot (such as the queue, skip tracker) that ensure that only one thread at a time can access and make changes to the queue.

So, in short, those commands aren't responding because they're waiting for another queue-related task to finish. In this case it looks like the downloader isn't downloading the YouTube video, keeping the queue locked.

You may not care about the things happening "behind-the-scenes" but I figured I'd explain it anyway. :)

Update `youtube-dl` and let me know if that helps at all.

**Comment by BackIsBachus** [comment]
No, thank you about the "behind-the-scene" part about the multi-threading and the semaphore! I kinda figured it may have something to do with that (given that all the commands interacted with the queue) but couldn't find the words at the moment and I wasn't sure if it was a reason or a consequence of the problem ^^

I installed `youtube-dl` from the Debian 8 repository yesterday evening so it is up to date. My current version is `2014.08.05` is it enough or do I need to install a newer version directly from youtube-dl website?
(I'll try that during the day anyway just to be sure)

**Comment by matthieugrieger** [comment]
Yeah you'll want to get a newer version. That one is pretty outdated.

Generally for `youtube-dl` it is best to install it manually with the script from the `youtube-dl` website.

**Comment by BackIsBachus** [comment]
I removed the oudated `youtube-dl` installed the latest version from the website (`2016.08.19` can't be fresher ^^) tried the previous steps again but the same problem happens: after the first song nothing happens and all the command interacting with the queue do not respond.

I have python 2.7.9 installed (which is one of the compatible version that can be used by `youtube-dl`).

I'll try to install it on another server than the mumble server to test if it has not something to do with some legacy problem on this VM (1vCPU and 500MB of RAM).

**Comment by matthieugrieger** [comment]
Yeah, the specs shouldn't be an issue. I ran MumbleDJ along with a web server and a Mumble server on a 512MB RAM server for quite awhile and never had any issues.

What happens if you run the following command? This is what MumbleDJ is running in the background:
```
youtube-dl --verbose --no-mtime --output test.video --format bestaudio --prefer-ffmpeg <youtube_url_here>
```

**Comment by BackIsBachus** [comment]
The only reason I had to up the RAM a bit was to compile the docker image but except this there was no problem indeed.
I'm using avconv right now, is the `--prefer-ffmpeg` option mandatory (because ffmpeg would work better maybe)?

When I run your command with `--prefer-avconv` (the default option normally) I get the following output
```
[debug] System config: []
[debug] User config: []
[debug] Command-line args: [u'--verbose', u'--no-mtime', u'--output', u'test.video', u'--format', u'bestaudio', u'--prefer-avconv', u'https://www.youtube.com/watch?v=6uR15cqTPJA']
[debug] Encodings: locale ANSI_X3.4-1968, fs ANSI_X3.4-1968, out ANSI_X3.4-1968, pref ANSI_X3.4-1968
[debug] youtube-dl version 2016.08.19
[debug] Python version 2.7.9 - Linux-3.16.0-4-amd64-x86_64-with-debian-8.5
[debug] exe versions: avconv 11.7-6, avprobe 11.7-6
[debug] Proxy map: {}
[youtube] 6uR15cqTPJA: Downloading webpage
[youtube] 6uR15cqTPJA: Downloading video info webpage
[youtube] 6uR15cqTPJA: Extracting video information
[youtube] 6uR15cqTPJA: Downloading MPD manifest
[debug] Invoking downloader on u'https://r1---sn-gxo5uxg-jqbe.googlevideo.com/videoplayback?id=eae475e5ca933c90&itag=171&source=youtube&requiressl=yes&mn=sn-gxo5uxg-jqbe&mm=31&ms=au&pl=21&mv=m&initcwndbps=19185000&pcm2cms=yes&ratebypass=yes&mime=audio/webm&gir=yes&clen=2324642&lmt=1399187496580863&dur=150.475&mt=1471722525&sver=3&signature=6D656B43BD801C30CCBC97F06EB11489E267CD40.26216A4C5E32D3294F4381B66E3951406798EB32&key=dg_yt0&upn=2rKqFrlHi5A&fexp=9419452,9422596,9428398,9431012,9433096,9433223,9433946,9434179,9435526,9435914,9436048,9437141,9438227,9438327,9438662,9438829,9439580,9440302,9441082,9441772,9442173,9442424,9442426,9442779,9444126,9444649,9444659,9445077,9445372&ip=160.228.158.56&ipbits=0&expire=1471744924&sparams=ip,ipbits,expire,id,itag,source,requiressl,mn,mm,ms,pl,mv,initcwndbps,pcm2cms,ratebypass,mime,gir,clen,lmt,dur'
[download] Destination: test.video
[download] 100% of 2.22MiB in 00:00
```
If I try to read the file with VLC it works just fine and the audio is perfect.

My problem is mainly getting from the 1st song to the 2nd: the 1st song I add to the playlist on MumbleDJ plays without any problem (I can hear it on Mumble), but it's when MumbleDJ tries to get to the 2nd that it seems to get stuck while holding the semaphore of the queue.

**Comment by matthieugrieger** [comment]
It doesn't matter if you use `--prefer-ffmpeg` or `--prefer-avconv`. They're essentially the same, `avconv` is just a fork of `ffmpeg`.

The download always being 2.22MiB in size does seem like an issue...

I'll execute the same command and post what I get.

**Comment by BackIsBachus** [comment]
My server does not uses IPv6, it is hosted on one of my association servers and we are our own small ISP for the campus (like we have a fiber coming in and have an IP range we can use but after that it's up to us to deal with it). The server is however behind a NAT and a firewall that can be pretty strict but I have access to it and can change stuff if needed (but for this server standard ports are open like 22, 80, 443, default mumble port).

I tried the command on a VPS hosted by OVH and `youtube-dl` now downloads the same 529.67MiB as you did. I'm gonna try run the bot from here to see if the problem persists from there.

**Comment by matthieugrieger** [comment]
Sounds good, let me know what you find out!

**Comment by BackIsBachus** [comment]
Well, the same problem still persists.
I also figured out the 2.22MiB problem: I forgot to delete `test.video` between 2 tests and I think `youtube-dl` assumed that the download was already complete. When I delete `test.video` between each try I normally download the 529.67MiB.
So that was more a problem with how I use `youtube-dl`.

If I activate the cache I find files with the right size in `.cache/mumbledj/` so I think the `youtube-dl` command runs without any problem.

**Comment by matthieugrieger** [comment]
Ah, that makes sense. There may be something wrong with MumbleDJ then. I'll have to see if I can reproduce the issue.

**Comment by AConsolePeasant** [comment]
I'm getting the same issue, however I removed my install because of it :\

**Comment by matthieugrieger** [comment]
Good news, looks like it is a bug that I introduced fairly recently. I'm experiencing it too (my server was running an older version of MumbleDJ).

Working on a fix.

**Comment by matthieugrieger** [comment]
@BackIsBachus, @AConsolePeasant: The issue should be fixed now. Looks like the last update I made accidentally created a deadlock.

**Comment by BackIsBachus** [comment]
Works like a charm now! Thank you for your time and awesome work!

On a side note do you mind if I add a few wiki pages with a few pointers and example of complete setups (for example add a systemd daemon script, run the bot as a dedicated user and related things) for less experimented Linux CLI users?

**Comment by matthieugrieger** [comment]
Go for it! That sounds great.

**Comment by BackIsBachus** [comment]
Indeed I'm referring to the Github wiki pages of this project.
